hask/02.hs

```diff
@@ -20,25 +20,20 @@ main = do
     print $ solve part1 contents
     print $ solve part2 contents
 
-solve f = sum . map (line f) . zip [1..] . lines
+solve f = sum . map f . zip [1..] . lines
 
-line :: (Eq a, Eq b, Num a) => ((a,b) -> Maybe a) -> (a,b) -> a
-line f s | res == Nothing = 0
-         | otherwise = fromJust res
-    where res = f s
+validColor c x =
+    last s /= c ||
+    read (head s) <= getMaxColor c
+  where s = words x
 
-validColor c x | (last s == c) = ((read $ head s) <= getMaxColor c)
-           | otherwise = True
-           where s = words $ x
+validColors s = and $
+    validColor <$> rgb <*> splitOn "," s
+  where rgb = words "red green blue"
```

Comment:
This one is cool. We use the `Applicative` to create all the instances in one line of code!